### PR TITLE
docs: update P2P readme

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Changes

- Introduction: replace comma with semicolon - separate but related statements joined by "however"
- §II.1 ¶5: add missing serial comma and move "and" to precede the final list item
- §II.4 ¶5: Assuming my interpretation is correct: - Fix / complete logic description for right neighbor - Reduce logic description by moving "both their immediate left and right neighbor" to precede the colon
- §II.5 ¶2: Rephrase for clarity
- §II.5.1 ¶2: add missing "or" and rewrite `(addresse(s))` as `(address(es))` - the "e" is part of the optional pluralization
- §II.5.2 ¶2: fix typo in "peer X's"
- §II.5 Membership Pings: add missing section number, §II.5.4
- §II.5.4 ¶1: add semicolon and rearrange "however" and comma to connect related statements, remove unnecessary "in", and reword "separation of raintree" as "separation from raintree"
- §II.6.1: Add missing period and convert to ordered list
- §II.6.2 Step 5: clarify that the shared secret key is **NEVER** transmitted as part of any message

### CODE REVIEW

**Comment by bryanchriswhite** [comment]
I had some additional questions / observations about areas that seemed to me like they could benefit from additional clarity but wasn't prepared to take a stab at it myself just yet. Maybe some of these belong in #only-stupid-questions (perhaps I'll link this comment there). 


### §II.1

- "This traversal algorithm follows a tree reconstruction algorithm such that the resulting tree is one of three branches and not two."
- "To determine Node S' layer" - what does "S' layer" mean? This phrase never re-used.
- "Count the exponents of 3 in the list..." - tough to explain, an example or viz would help
- What entails a valid node ID
    - What does the address space look like?
    - What **is** the distance metric?
- Where does `1.5` come from in "left branch target = Node position + targetListSize / 1.5"

### §II.2

- RainTree Gossip Algorithm
    - Should `S` be received as an argument to `Gossip`?
    - What's the relationship between `S` and `self`?

### §II.3

- peudo-code for IGYW doesn't cover the non-response scenario

### §II.6.4

- "We will rely on TCP/IP with a handshake for direct communications and Gossip where as we intend to use UDP for churn management communication."

### §II.6.6

- "Peer connections could be encrypted..." - **could** be? Maybe we want to be a bit firmer / committal here?

**Comment by bryanchriswhite** [comment]
@Olshansk, I just discovered your comment regarding the [p2p/v102](https://github.com/pokt-network/pocket-network-protocol/tree/p2p/v102) branch while reading #2.

**Review by Olshansk — APPROVED** [review]

**Comment by Olshansk** [comment]
> I had some additional questions / observations about areas that seemed to me like they could benefit from additional clarity but wasn't prepared to take a stab at it myself just yet. Maybe some of these belong in #only-stupid-questions (perhaps I'll link this comment there).

Very much expected and makes sense. We can start here and take it into an offline chat if need be.

> I just discovered your comment regarding the [p2p/v102](https://github.com/pokt-network/pocket-network-protocol/tree/p2p/v102?rgh-link-date=2023-01-19T13%3A59%3A52Z) branch while reading https://github.com/pokt-network/pocket-network-protocol/pull/2.

That branch is quite stale at this point but I can do a quick re-review and merge it in on top of your changes. Hopefully it serves to show that these specs are just guidelines and it's put to us to define the path forward, similar to what I submitted earlier this week in #25.

## §II.1

> "This traversal algorithm follows a tree reconstruction algorithm such that the resulting tree is one of three branches and not two." - I'm not sure what I'm supposed to take away from this, I also don't understand what is meant by "three branches".

A ternary tree. Like a binary tree (2 children) but where each node has three children.

> "To determine Node S' layer" - what does "S' layer" mean? This phrase never re-used.

The `Source` node. A.k.a the originator of the message that starts the gossip. The root of the tree.

> "Count the exponents of 3 in the list..." - tough to explain, an example or viz would help

Hopefully when you make your way to the python simulator (https://github.com/pokt-network/rain-tree-sim/tree/main/python), it'll make more sense.

> What entails a valid node ID

RainTree (structured) gossip disseminates messages between **staked** actors, as per the world state. We need to efficiently disseminate the message throughout this set of nodes. 

Full nodes (non-staked) will use random gossip since their latency to reach a synched state is a lower priority.

> What does the address space look like?

I don't have a direct answer on how to visualise the address space, but we use a transformation of the Ed25519 publickKey to generate addresses:

```go
func (pub Ed25519PublicKey) Address() Address {
	hash := sha256.Sum256(pub[:])
	return hash[:AddressLen]
}
```

Is that what you were looking for?

> What is the distance metric?

I do not believe `distance metric` is used correctly. We sort the addresses of staked actors lexicographically. 

> Where does 1.5 come from in "left branch target = Node position + targetListSize / 1.5"

I believe the python and go code for this are better methods of understanding the implementation. I did not prioritize going back to update the specification yet, unfortunately.

See [addrbook_utils.go#L10](https://github.com/pokt-network/pocket/blob/main/p2p/raintree/addrbook_utils.go#L10) in the source code:

```go
const (
	firstMsgTargetPercentage  = float64(1) / float64(3)
	secondMsgTargetPercentage = float64(2) / float64(3)
	shrinkagePercentage       = float64(2) / float64(3)
	maxLevelsLogBase          = float64(3)
	floatPrecision            = float64(0.0000001)
)
```

Or [class RainTreeConfigs](https://github.com/pokt-network/rain-tree-sim/blob/main/python/simulator_types.py#L11) in the simulator:

```python
@dataclass
class RainTreeConfigs:
    num_nodes: int
    t1_per: float  # % of location in addr_book for the node's 1st target (whom to send a message to)
    t2_per: float  # % of location in addr_book for the node's 1st target (whom to send a message to)
    shrinkage_per: float  # addr book shrinkage coefficient (i.e. % of addr book that should be kept for the next level)
    max_theoretical_depth: int
    addr_book: List[str]
```

## §II.2

> Should S be received as an argument to Gossip?

Yes, it is the source node that starts gossiping.

Screenshot 2023-01-19 at 8 20 32 PM


> What's the relationship between S and self?

Self is the node propagating the message at every height, while `S` is the originator.

## §II.3

> peudo-code for IGYW doesn't cover the non-response scenario

Noted. Revisiting the cleanup & redundancy layer is something we will need to do.

The ternary structured gossip algorithm is great for disseminating messages quickly to active actors but there are still some open questions to resolve.

## §II.6.4

> "We will rely on TCP/IP with a handshake for direct communications and Gossip where as we intend to use UDP for churn management communication."

Agreed that this is not clear. Whether it's a send or a gossip propagation, I believe we should be using TCP everywhere.

## §II.6.6

> "Peer connections could be encrypted..." - could be? Maybe we want to be a bit firmer / committal here?

The tl;dr is "TCP" and avoid doing anything innovative. I believe we can remove this section altogether and just use what LibP2P gives us for free out of the box.

**Comment by Olshansk** [comment]
@bryanchriswhite Great questions. I rushed through the answers slightly so let me know if you want to hop on a call and discuss this in greater detail.

**Comment by bryanchriswhite** [comment]
@Olshansk no worries, thanks for the quick turnaround! The additional detail did clarify most points :pray:, I agree that it probably makes sense to switch to a higher bandwidth medium for the rest.
